airflow/example_dags/example_workday_timetable.py

```diff
@@ -0,0 +1,92 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Example DAG demostrating how to implement a custom timetable for a DAG."""
+
+# [START howto_timetable]
+from datetime import timedelta
+from typing import Optional
+
+from pendulum import Date, DateTime, Time, timezone
+
+from airflow import DAG
+from airflow.operators.dummy import DummyOperator
+from airflow.timetables.base import DagRunInfo, DataInterval, TimeRestriction, Timetable
+
+UTC = timezone("UTC")
+
+
+class AfterWorkdayTimetable(Timetable):
+
+    # [START howto_timetable_infer_data_interval]
+    def infer_data_interval(self, run_after: DateTime) -> DataInterval:
+        weekday = run_after.weekday()
+        if weekday in (0, 6):  # Monday and Sunday -- interval is last Friday.
+            days_since_friday = (run_after.weekday() - 4) % 7
+            delta = timedelta(days=days_since_friday)
+        else:  # Otherwise the interval is yesterday.
+            delta = timedelta(days=1)
+        start = DateTime.combine((run_after - delta).date(), Time.min).replace(tzinfo=UTC)
+        return DataInterval(start=start, end=(start + timedelta(days=1)))
+
+    # [END howto_timetable_infer_data_interval]
+
+    # [START howto_timetable_next_dagrun_info]
+    def next_dagrun_info(
+        self,
+        last_automated_dagrun: Optional[DateTime],
+        restriction: TimeRestriction,
+    ) -> Optional[DagRunInfo]:
+        if last_automated_dagrun is not None:
+            # There was a previous run on the regular schedule.
+            # last_automated_dagrun os the last run's logical date.
+            weekday = last_automated_dagrun.weekday()
+            if 0 <= weekday < 4:  # Monday through Thursday -- next is tomorrow.
+                delta = timedelta(days=1)
+            else:  # Week is ending -- skip to next Monday.
+                delta = timedelta(days=(7 - weekday))
+            start = DateTime.combine((last_automated_dagrun + delta).date(), Time.min)
```

Comment:
Reminder @uranusjr that there is a bug in this example. The next interval should start at a delta from the start of the incoming dagrun, not the dagrun execution time.

Reply:
Some additional context:

> it takes the execution time (which falls after the interval ends) and then adds 1 day to it to find the next start of the next interval. But that ends up skipping an interval, if I'm not mistaken.

This makes me think maybe we should pass the entire `DagRunInfo` into `next_dagrun_info` instead of just the previous run date. It is of course not too difficult to reverse-infer the data interval from the run date in this example, but it might not be always as easy or even possible in more elaborated data interval logic.

Reply:
Yeah this sounds like a good idea

Reply:
Hmm, on a closer look, the current logic seems to be correct? The default behaviour of `DagRunInfo` actually defines the execution date as the _start_ of the interval, so I think using `execution_date + delta` is correct. See the `logical_date` property defined here:

https://github.com/apache/airflow/blob/873d4d1f1aad214b36218d80000f5175ae284f6f/airflow/timetables/base.py#L56-L91

Reply:
Regarding also passing in the data interval. This turns out to be more work than I originally anticipated because it means we’ll need to deprecate `DAG.following_schedule()` (which operates on `execution_date` only).  I’ll work on it in the following days.

Reply:
New `next_dagrun_info` implementation using data interval instead was merged in #18088. I’ve rewritten the example to match the actual implementation.